### DISCUSSION TITLE
Image regression testing/debugging on Windows

### DISCUSSION THREAD

**Discussion by user27182** [pyvista/pyvista#4698] [Q&A]
I am trying to resolve some failed image regression tests. I have looked at the [notes regarding image regression testing](https://docs.pyvista.org/version/stable/extras/developer_notes.html#notes-regarding-image-regression-testing) which says this is only done on Linux, but I am running Windows locally. I have a few questions:

1. The linked URL from the docs [pytest-pyvista](https://pytest.pyvista.org) is broken and so I am unable to get more info about this there. Does anyone know if there is a working link somewhere?
2. Is it not possible to debug failed CI tests (e.g. from a pull request) on Windows?
3. If it _is_ possible, I am running into problems executing, for example, the command `pytest tests/plotting/ -k volume --generated_image_dir debug_images`, which fails to run and gives me the error:
`INTERNALERROR> importlib.metadata.PackageNotFoundError: No package metadata was found for pyvista`
Trying to resolve this with `pip install metadata` also fails, and generates the following:
```
ERROR: Cannot install metadata==0.1, metadata==0.1.1 and metadata==0.2 because these package versions have conflicting dependencies.

The conflict is caused by:
    metadata 0.2 depends on hachoir-core==1.3.3
    metadata 0.1.1 depends on hachoir-core==1.3.3
    metadata 0.1 depends on hachoir-core==1.3.3
```
Running `pip install pytest-pyvista` and `pip install hachoir` does not seem to resolve the `importlib.metadata.PackageNotFoundError` error either.

Any ideas on how to proceed?

Ultimately, I need to generate a number of debug images for review and replace existing test images with new ones to include as part of a PR. If this is not something that can be or should be done on Windows (e.g. in case it's preferred/required to have Linux-generated test images), please let me know. Thanks!

System info from `pv.Report()`:
```
                OS : Windows
            CPU(s) : 12
           Machine : AMD64
      Architecture : 64bit
               RAM : 31.8 GiB
       Environment : IPython
        GPU Vendor : Intel
      GPU Renderer : Intel(R) UHD Graphics 630
       GPU Version : 4.5.0 - Build 30.0.100.9864
  MathText Support : False
  Python 3.10.2 (tags/v3.10.2:a58ebcc, Jan 17 2022, 14:12:15) [MSC v.1929 64
  bit (AMD64)]
           pyvista : 0.42.dev0
               vtk : 9.2.6
             numpy : 1.25.1
        matplotlib : 3.7.2
            scooby : 0.7.2
             pooch : v1.7.0
           imageio : 2.31.1
           IPython : 8.14.0
          colorcet : 3.0.1
           cmocean : v3.0.3
        ipyvtklink : 0.2.3
        ipywidgets : 7.7.5
             scipy : 1.11.1
              tqdm : 4.65.0
            meshio : 5.3.4
        jupyterlab : 4.0.3
         pythreejs : 2.4.2
    pytest_pyvista : 0.1.8
             trame : 2.5.0
      trame_client : 2.9.4
      trame_server : 2.11.4
         trame_vtk : 2.5.4
      nest_asyncio : 1.5.6
```

**Comment by MatthewFlamm** [comment]
I have never personally used windows to develop with pyvista.  But the CI does run image regression tests on windows.

Can you provide a traceback for `INTERNALERROR> importlib.metadata.PackageNotFoundError: No package metadata was found for pyvista`.  Note this error means that it cannot find pyvista, it isn't looking for a package called `metadata`.

This is weird since you can get the report.  This makes me think your pyvista and pytest environments aren't the same.

**Reply by user27182** [reply]
Thanks @MatthewFlamm for your suggestion. The `PackageNotFoundError` was generated when trying to execute the tests from an external shell with a `pyvista` venv activated, perhaps this was the source of `pyvista` not being found. I wasn't able to resolve the `PackageNotFoundError`, and still can't run the tests from an external shell, but I was able to execute the plotting tests by creating a test configuration within PyCharm and running the configuration from within the IDE. I was able to save the test images to `debug_images` and review them / compare them to the caches images accordingly. Executing the tests within the IDE also made it much easier to debug failed tests. So, my problem has been resolved. Thanks!

**Reply by MatthewFlamm** [reply]
Sounds like you have some python environment issue possibly exacerbated by windows.  Glad it was resolved.  I still suspect you might have pytest being invoked from a different environment (maybe even pyvista import could be this way given that pyvista doesn't use a 'src-layout' model for packaging).  Python is tricky this way.
